File: retailedge/retailedge/report/retailedge_edgepay_reconciliation_readiness/retailedge_edgepay_reconciliation_readiness.py
# -*- coding: utf-8 -*-
from __future__ import annotations
import frappe
from frappe import _
from frappe.utils import get_first_day, getdate, nowdate
from retailedge.services.edgepay_reconciliation_readiness import (
	get_edgepay_reconciliation_readiness,
	find_edgepay_payment_entry_bank_match_candidates
)

def execute(filters=None):
	filters = frappe._dict(filters or {})
	filters.setdefault("from_date", str(get_first_day(nowdate())))
	filters.setdefault("to_date", str(nowdate()))
	
	# Fetch evidence records matching filters
	db_filters = {}
	if filters.get("from_date") and filters.get("to_date"):
		db_filters["creation"] = ["between", [filters.get("from_date") + " 00:00:00", filters.get("to_date") + " 23:59:59"]]
	if filters.get("review_status"):
		db_filters["review_status"] = filters.get("review_status")
	if filters.get("reconciliation_status"):
		db_filters["reconciliation_status"] = filters.get("reconciliation_status")
		
	evidences = frappe.get_all("RetailEdge EdgePay Payment Evidence", filters=db_filters, fields=[
		"name", "source_doctype", "source_name", "amount", "currency", 
		"provider_reference", "payment_entry", "submission_status", 
		"reconciliation_status", "reconciliation_message"
	])
	
	rows = []
	for ev in evidences:
		# Double check if filter by company matches (if filter company is set and source doc supports company)
		if filters.get("company"):
			source_company = frappe.db.get_value(ev.source_doctype, ev.source_name, "company")
			if source_company != filters.get("company"):
				continue
				
		# Get candidate count
		candidates = find_edgepay_payment_entry_bank_match_candidates(ev.name)
		
		# Resolve blocking reason (which is reconciliation_message)
		blocking_reason = ev.reconciliation_message if ev.reconciliation_status in ("Blocked", "Exception") else ""
		
		rows.append({
			"evidence": ev.name,
			"source_doctype": ev.source_doctype,
			"source_name": ev.source_name,
			"amount": ev.amount,
			"currency": ev.currency,
			"provider_reference": ev.provider_reference,
			"payment_entry": ev.payment_entry,
			"submission_status": ev.submission_status,
			"reconciliation_status": ev.reconciliation_status,
			"candidate_count": len(candidates),
			"blocking_reason": blocking_reason
		})
		
	return get_columns(), rows, None, None, get_report_summary(rows)

def get_columns():
	return [
		{"label": _("Payment Evidence"), "fieldname": "evidence", "fieldtype": "Link", "options": "RetailEdge EdgePay Payment Evidence", "width": 170},
		{"label": _("Source DocType"), "fieldname": "source_doctype", "fieldtype": "Data", "width": 130},
		{"label": _("Source Name"), "fieldname": "source_name", "fieldtype": "Dynamic Link", "options": "source_doctype", "width": 160},
		{"label": _("Amount"), "fieldname": "amount", "fieldtype": "Currency", "width": 110},
		{"label": _("Currency"), "fieldname": "currency", "fieldtype": "Data", "width": 90},
		{"label": _("Provider Reference"), "fieldname": "provider_reference", "fieldtype": "Data", "width": 150},
		{"label": _("Payment Entry"), "fieldname": "payment_entry", "fieldtype": "Link", "options": "Payment Entry", "width": 170},
		{"label": _("Submission Status"), "fieldname": "submission_status", "fieldtype": "Data", "width": 120},
		{"label": _("Reconciliation Status"), "fieldname": "reconciliation_status", "fieldtype": "Data", "width": 150},
		{"label": _("Candidate Count"), "fieldname": "candidate_count", "fieldtype": "Int", "width": 120},
		{"label": _("Blocking Reason"), "fieldname": "blocking_reason", "fieldtype": "Small Text", "width": 250}
	]

def get_report_summary(rows):
	return [
		{"label": _("Reconciled"), "value": sum(1 for row in rows if row.get("reconciliation_status") == "Reconciled"), "datatype": "Int", "indicator": "Green"},
		{"label": _("Ready"), "value": sum(1 for row in rows if row.get("reconciliation_status") == "Ready"), "datatype": "Int", "indicator": "Blue"},
		{"label": _("Matched"), "value": sum(1 for row in rows if row.get("reconciliation_status") == "Matched"), "datatype": "Int", "indicator": "Orange"},
		{"label": _("Blocked / Not Ready"), "value": sum(1 for row in rows if row.get("reconciliation_status") in ("Blocked", "Exception", "Not Ready")), "datatype": "Int", "indicator": "Red"}
	]

File: retailedge/retailedge/report/retailedge_edgepay_reconciliation_readiness/retailedge_edgepay_reconciliation_readiness.js
function configureOperationalReportRefresh(report) {
	if (!report || report.__retailedgeAutoRefreshConfigured) {
		return;
	}
	report.__retailedgeAutoRefreshConfigured = true;
	report.ignore_prepared_report = true;
	report.prepared_report = false;
	report.prepared_report_name = null;
	report.prepared_report_document = null;
	report.__retailedgeAutoRefreshReady = true;
	(report.filters || []).forEach((filter) => {
		const originalOnChange = filter.on_change;
		filter.on_change = function (queryReport) {
			if (typeof originalOnChange === "function") {
				originalOnChange.call(this, queryReport || report);
			}
			if (!report.__retailedgeAutoRefreshReady) {
				return;
			}
			scheduleOperationalReportRefresh(queryReport || report);
		};
	});
}

function scheduleOperationalReportRefresh(report) {
	if (!report) {
		return;
	}
	if (report.__retailedgeRefreshTimer) {
		clearTimeout(report.__retailedgeRefreshTimer);
	}
	report.__retailedgeRefreshTimer = setTimeout(() => {
		report.refresh();
	}, 200);
}

function forceOperationalPrimaryAction(report) {
	if (!report || !report.page || typeof report.page.set_primary_action !== "function") {
		return;
	}
	report.page.set_primary_action(__("Refresh Report"), () => {
		report.refresh();
	});
}

frappe.query_reports["RetailEdge EdgePay Reconciliation Readiness"] = {
	filters: [
		{ fieldname: "company", label: __("Company"), fieldtype: "Link", options: "Company" },
		{ fieldname: "from_date", label: __("Date From"), fieldtype: "Date", default: frappe.datetime.month_start(), reqd: 1 },
		{ fieldname: "to_date", label: __("Date To"), fieldtype: "Date", default: frappe.datetime.get_today(), reqd: 1 },
		{ fieldname: "review_status", label: __("Review Status"), fieldtype: "Select", options: "\nPending Review\nReviewed\nRejected\nException" },
		{ fieldname: "reconciliation_status", label: __("Reconciliation Status"), fieldtype: "Select", options: "\nNot Ready\nReady\nMatched\nReconciled\nBlocked\nException" }
	],
	onload(report) {
		configureOperationalReportRefresh(report);
		forceOperationalPrimaryAction(report);
		
		report.page.add_inner_button(__('Create Match Review'), () => {
			// Step 1: Prompt user for Evidence Name
			frappe.prompt([
				{
					label: __('Payment Evidence'),
					fieldname: 'evidence',
					fieldtype: 'Link',
					options: 'RetailEdge EdgePay Payment Evidence',
					get_query: () => {
						return {
							filters: {
								reconciliation_status: 'Ready',
								docstatus: 0
							}
						};
					},
					reqd: 1
				}
			], (values) => {
				let evidence_name = values.evidence;
				// Step 2: Fetch candidates
				frappe.call({
					method: 'retailedge.api.find_edgepay_payment_entry_bank_match_candidates',
					args: {
						evidence_name: evidence_name
					},
					callback: function(r) {
						let candidates = r.message || [];
						if (candidates.length === 0) {
							frappe.msgprint(__('No candidate Bank Transactions found matching this evidence.'));
							return;
						}
						
						let options = candidates.map(c => ({
							value: c.bank_transaction,
							label: `${c.bank_transaction} - Date: ${c.date} | Ref: ${c.reference_number || 'None'} | Conf: ${c.confidence}`
						}));
						
						frappe.prompt([
							{
								label: __('Bank Transaction Candidate'),
								fieldname: 'bank_transaction',
								fieldtype: 'Select',
								options: options,
								reqd: 1
							}
						], (cand_values) => {
							// Step 3: Call preflight first
							frappe.call({
								method: 'retailedge.api.get_edgepay_bank_match_review_preflight',
								args: {
									evidence_name: evidence_name,
									bank_transaction_name: cand_values.bank_transaction
								},
								callback: function(res) {
									if (res.message && res.message.ok) {
										// Step 4: Create match review
										frappe.call({
											method: 'retailedge.api.create_edgepay_bank_match_review',
											args: {
												evidence_name: evidence_name,
												bank_transaction_name: cand_values.bank_transaction
											},
											callback: function(create_res) {
												if (create_res.message && create_res.message.ok) {
													frappe.show_alert({
														message: __('Bank Match Review {0} created successfully.', [create_res.message.review_name]),
														indicator: 'green'
													});
													report.refresh();
												} else {
													frappe.msgprint(create_res.message.message || __('Failed to create match review.'));
												}
											}
										});
									} else {
										frappe.msgprint(res.message.message || __('Preflight validation failed.'));
									}
								}
							});
						}, __('Select Candidate'), __('Create Match Review'));
					}
				});
			}, __('Select Payment Evidence'), __('Next'));
		});

		report.page.add_inner_button(__('Confirm Match Review'), () => {
			frappe.prompt([
				{
					label: __('Payment Evidence'),
					fieldname: 'evidence',
					fieldtype: 'Link',
					options: 'RetailEdge EdgePay Payment Evidence',
					get_query: () => {
						return {
							filters: {
								reconciliation_status: 'Matched',
								docstatus: 0
							}
						};
					},
					reqd: 1
				}
			], (values) => {
				let evidence_name = values.evidence;
				frappe.db.get_doc('RetailEdge EdgePay Payment Evidence', evidence_name).then(doc => {
					let review_name = doc.linked_bank_match_review;
					if (!review_name) {
						frappe.msgprint(__('No linked Bank Match Review found on evidence.'));
						return;
					}
					frappe.call({
						method: 'retailedge.api.get_edgepay_bank_match_confirmation_preflight',
						args: {
							evidence_name: evidence_name,
							review_name: review_name
						},
						callback: function(res) {
							if (res.message && res.message.ok) {
								frappe.confirm(
									__('Are you sure you want to confirm this Bank Match Review?'),
									() => {
										frappe.call({
											method: 'retailedge.api.confirm_edgepay_bank_match_review',
											args: {
												evidence_name: evidence_name,
												review_name: review_name
											},
											callback: function(confirm_res) {
												if (confirm_res.message && confirm_res.message.ok) {
													frappe.show_alert({
														message: __('Bank Match Review confirmed successfully.'),
														indicator: 'green'
													});
													report.refresh();
												} else {
													frappe.msgprint(confirm_res.message.message || __('Failed to confirm match review.'));
												}
											}
										});
									}
								);
							} else {
								frappe.msgprint(res.message.message || __('Confirmation preflight validation failed.'));
							}
						}
					});
				});
			}, __('Select Payment Evidence to Confirm'), __('Confirm'));
		});
	},
	after_refresh(report) {
		forceOperationalPrimaryAction(report);
	}
};
